drivers: stepper: introduce stepper_ctrl.h for stepper motion control apis

jilaypandya opened the pull request:
~~Original PR~~
~~following the approach of transceiver.h in can driver subsystem, refactoring stepper_drv apis to stepper_drv.h~~

**First Commit**

- Dropped the `_drv` from the `stepper_drv_driver_api`. `stepper.h` only consists of `stepper driver` APIs.

- All the `stepper_motion_control` APIs have been moved to `include/zephyr/drivers/stepper/stepper_ctrl.h`.

@carlescufi @henrikbrixandersen would this be okay for now?

henrikbrixandersen reviewed (changes requested):
This is nothing like what the CAN driver subsystem does (or any other driver subsystem for that matter). 

What's the rationale behind this?

jilaypandya commented:
Ahh okay, i found the recent split that we introduced in Stepper driver subsystem i.e. Stepper (Stepper Controller) and Stepper-DRV (Stepper driver) highly analogous to how CAN Controller and CAN Transceiver APIs are structured. But maybe i am missing out on something :)

jilaypandya committed:
drivers: stepper: split stepper.h and rename apis

rename stepper_drv_driver_api to stepper_driver_api
rename stepper_driver_api to stepper_ctrl_driver_api

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

jilaypandya commented:
- So `stepper_motion_controller_driver_api` is too long. What we could do maybe is `stepper_mc_driver_api`.
- `stepper_controller_driver_api` might be ambiguous.
- `stepper_motion_driver_api` would be another option.

I went for `stepper_motor_driver_api` as I saw something [similiar in one of the vendor sdks](https://renesas.github.io/fsp/group___m_o_t_o_r___a_p_i.html#structmotor__api__t).

jilaypandya commented:
Okay, so  we have some consensus now on `stepper_mc_driver_api`, great :). 
Subsequently using `drv` in `stepper_drv_driver_api` is not really required anymore. 
Looking at the following image, i think it's pretty evident as to what constitutes a `stepper_driver` and a `stepper_motion_controller_driver`:
<img width="399" height="234" alt="Image" src="https://github.com/user-attachments/assets/bce38e92-531d-47af-bc2c-83f56ef04bf7" style="background-color: white;"  />

Additionally it's been brought to my attention that using `_drv` is not something that we do or should do in tree.

jbehrensnx commented:
Clear yes, but technically wrong, as we are talking about drivers for stepper drivers. Is `_mc` actually ok if `_drv` is an issue, as both are abbreviations?

jbehrensnx commented on doc/hardware/peripherals/stepper/index.rst:
```diff
@@ -5,32 +5,32 @@ Steppers
-- **Move by** +/- micro-steps also known as **relative movement** using :c:func:`stepper_move_by`.
-- **Move to** a specific position also known as **absolute movement** using :c:func:`stepper_move_to`.
+- Configure **reference position** in microsteps using :c:func:`stepper_ctrl_set_reference_position`
+  and :c:func:`stepper_ctrl_get_actual_position`.
```
Admittedly slightly out of scope, but as a lot of renaming is happening anyway: Changing function names so that the getter and setter functions for the position have a consistent name.

jbehrensnx commented on doc/hardware/peripherals/stepper/individual_controller_driver.rst:
```diff
@@ -3,8 +3,8 @@
 Individual Stepper Motion Controller and Driver
```
Possibly drop `Motion` from Motion Controller, as the API is only called Stepper Controller. This applies to all occurrences in this file.

jilaypandya replied:
It's ofcourse an api for stepper motion controllers. I would really like to focus on renaming the api names as of now. Documentation can always be improved.

jbehrensnx commented on drivers/stepper/gpio_stepper/gpio_step_dir.c:
```diff
@@ -3,22 +3,22 @@
-LOG_MODULE_REGISTER(gpio_step_dir_stepper, CONFIG_STEPPER_LOG_LEVEL);
+LOG_MODULE_REGISTER(gpio_step_dir, CONFIG_STEPPER_LOG_LEVEL);
 
 struct zephyr_gpio_step_dir_stepper_config {
```
See comment above concerning a consistent naming scheme.

jbehrensnx commented on drivers/stepper/Kconfig.fake:
```diff
@@ -4,8 +4,8 @@
 
 config FAKE_STEPPER
-	bool "Fake stepper and stepper_drv driver"
+	bool "Fake stepper motor controller and stepper driver"
```
```suggestion
	bool "Fake stepper controller and stepper driver"
```

jbehrensnx reviewed (changes requested):
A lot of this is minor stuff, where things have not been correctly updated.

There are however also some points that I think are worth discussing, mainly `stepper_ctrl` vs `stepper_controller` for the API name.

That said, there are in of itself no majore blocking issues from my side.

jilaypandya commented on include/zephyr/drivers/stepper/stepper_ctrl.h:
```diff
@@ -0,0 +1,408 @@
+ */
+
+/**
+ * @file drivers/stepper/stepper_ctrl.h
```
The shorthand in general is not an issue for what i've understood, the issue is specifically with "drv". https://github.com/zephyrproject-rtos/zephyr/issues/101330#issuecomment-3878294704

mbolivar reviewed (commented):
No objections to the DT changes, but otherwise I'm not able to review this one, sorry.

jbehrensnx commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -656,43 +656,50 @@ Stepper
-
-* :dtcompatible:`zephyr,gpio-step-dir-stepper` implements :c:group:`stepper_interface` for
+* Since :github:`91979` & lateron :github:`101001`,
+  following APIs have been moved from :c:group:`stepper_driver_interface` to :c:group:`stepper_ctrl_interface`:
```
```suggestion
  the following functions have been moved from :c:group:`stepper_driver_interface` to :c:group:`stepper_ctrl_interface`:
```

jbehrensnx commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -656,43 +656,50 @@ Stepper
+  * :c:func:`stepper_get_actual_position` is replaced by :c:func:`stepper_ctrl_get_actual_position`.
+  * :c:func:`stepper_set_microstep_interval` is replaced by :c:func:`stepper_ctrl_set_microstep_interval`.
+
+* Following events have been from :c:enum:`stepper_event` to :c:enum:`stepper_ctrl_event`:
```
```suggestion
* The following events have been moved from :c:enum:`stepper_event` to :c:enum:`stepper_ctrl_event`:
```

jbehrensnx commented on include/zephyr/drivers/stepper/stepper.h:
```diff
@@ -0,0 +1,276 @@
+#endif
+
+/**
+ * @brief Interfaces for stepper motor controllers and driver.
```
I think this first block should be removed, but I am not sure.

jbehrensnx commented on include/zephyr/drivers/stepper/stepper_ctrl.h:
```diff
@@ -0,0 +1,408 @@
+/**
+ * @file drivers/stepper/stepper_ctrl.h
+ * @ingroup stepper_ctrl_interface
+ * @brief Main header file for stepper controller driver API.
```
```suggestion
 * @brief Main header file for stepper motion controller driver API.
```

Other instances in this file as well.

smaherdev reviewed (dismissed):
Some minor suggestions for consistency of naming

bjarki-andreasen commented:
> who should be the assignee for this PR?

You, its entirely stepper is it not?

JarmouniA commented on include/zephyr/drivers/stepper/stepper.h:
```diff
@@ -0,0 +1,276 @@
+#endif
+
+/**
+ * @brief Interfaces for stepper motion controllers and drivers.
```
please use "controller" exclusively to refer to hardware, instead of "driver" to avoid confusion with software "driver".

jilaypandya replied:
I mean that's the whole issue over here. There is no clear alternative to the word driver when it comes to stepper motors. Have tried to explain this in the linked RFC as well.
<img width="382"  alt="image" src="https://github.com/user-attachments/assets/8104a7d5-f1a7-48b1-964e-eae8b08468d9" />

JarmouniA replied:
Ok, a compromise would be to always add "HW" or "hardware" before "driver" when referring to the hardware block.

JarmouniA reviewed (changes requested):
Put DT-related renaming in a separate commit.

JarmouniA commented on doc/hardware/peripherals/stepper/individual_controller_driver.rst:
```diff
@@ -14,25 +14,25 @@ controller:
 
-        /* DEVICE_API: stepper api */
+        /* DEVICE_API: stepper_ctrl_driver_api */
         step_dir_motion_control: step_dir_motion_control {
```
https://devicetree-specification.readthedocs.io/en/latest/chapter2-devicetree-basics.html#node-names

jilaypandya replied:
Will amend this. Thanks :)

jilaypandya committed:
drivers: stepper: rename dt-bindings

stepper-ctrl is used as suffix for motion control drivers
stepper-driver is used as suffix for stepper drivers in mfd devices

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

JarmouniA commented on doc/hardware/peripherals/stepper/integrated_controller_driver.rst:
```diff
@@ -4,17 +4,17 @@ Integrated Stepper Motion Control and Driver
+have to be modelled as multi-functional-device in device tree, implementing both :c:group:`stepper_ctrl_interface`
+and :c:group:`stepper_driver_interface` APIs. An example of such a device is :dtcompatible:`adi,tmc50xx`.
+:c:group:`stepper_driver_interface` API is implemented by :dtcompatible:`adi,tmc50xx-stepper-driver` and
+:c:group:`stepper_ctrl_interface` API is implemented by :dtcompatible:`adi,tmc50xx-stepper-ctrl`.
```
It is one device but with multiple functional blocks.
My issue is with the wording "....API is implemented by :dtcompatible:...", bindings represent hardware, they do not implement software APIs, software drivers do.

jilaypandya committed:
doc: stepper: address api & dt-bindings renaming

address api and dt-bindings related changes in stepper
documentation and migration-guide.

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

jbehrensnx reviewed (dismissed):
From my side, no issues remain, that definitely need to be resolved in this PR

jilaypandya commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -656,43 +656,50 @@ Stepper
-
-* :dtcompatible:`zephyr,gpio-step-dir-stepper` implements :c:group:`stepper_interface` for
+* Since :github:`91979` & lateron :github:`101001`,
+  the following functions have been moved from :c:group:`stepper_driver_interface` to :c:group:`stepper_ctrl_interface`:
```
done

JarmouniA commented on doc/hardware/peripherals/stepper/integrated_controller_driver.rst:
```diff
@@ -3,18 +3,16 @@
-:c:group:`stepper_drv_interface` API is implemented by :dtcompatible:`adi,tmc50xx-stepper-drv`.
+Devices which comprise of both motion controller and a stepper driver functions in a single IC. These devices
+have to be modelled as multi-functional-device in device tree, implementing both :c:group:`stepper_ctrl_interface`
+and :c:group:`stepper_hw_driver_interface` APIs. An example of such a device is :dtcompatible:`adi,tmc50xx`.
```
```suggestion
A device comprised of both motion controller and stepper driver blocks in a single IC
is modeled as a multi-functional device in devicetree, and has two software drivers implementing :c:group:`stepper_ctrl_interface`
and :c:group:`stepper_hw_driver_interface` APIs respectively. An example of such device is :dtcompatible:`adi,tmc50xx`.
```

JarmouniA commented on doc/hardware/peripherals/stepper/integrated_controller_driver.rst:
```diff
@@ -32,19 +30,17 @@ and :c:group:`stepper_drv_interface` APIs. An example of such a device is :dtcom
 
-           /* DEVICE_API: stepper_drv api */
-           tmc50xx_0_stepper_driver: tmc50xx_0_stepper_driver {
+           tmc50xx_0_stepper_driver: tmc50xx-0-stepper_driver {
```
```suggestion
           tmc50xx_0_stepper_driver: stepper-driver-0 {
```

JarmouniA commented on doc/hardware/peripherals/stepper/integrated_controller_driver.rst:
```diff
@@ -32,19 +30,17 @@ and :c:group:`stepper_drv_interface` APIs. An example of such a device is :dtcom
 
-           /* DEVICE_API: stepper api */
-           tmc50xx_0_motion_controller: tmc50xx_0_motion_controller {
+           tmc50xx_0_motion_controller: tmc50xx-0-motion_controller {
```
```suggestion
           tmc50xx_0_motion_controller: motion-controller-0 {
```

JarmouniA commented on doc/hardware/peripherals/stepper/integrated_controller_driver.rst:
```diff
@@ -53,19 +49,17 @@ and :c:group:`stepper_drv_interface` APIs. An example of such a device is :dtcom
 
-           /* DEVICE_API: stepper_drv api */
-           tmc50xx_1_stepper_driver: tmc50xx_1_stepper_driver {
+           tmc50xx_1_stepper_driver: tmc50xx-1-stepper_driver {
```
```suggestion
           tmc50xx_1_stepper_driver: stepper-driver-1 {
```

JarmouniA commented on doc/hardware/peripherals/stepper/integrated_controller_driver.rst:
```diff
@@ -53,19 +49,17 @@ and :c:group:`stepper_drv_interface` APIs. An example of such a device is :dtcom
 
-           /* DEVICE_API: stepper api */
-           tmc50xx_1_motion_controller: tmc50xx_1_motion_controller {
+           tmc50xx_1_motion_controller: tmc50xx-1-motion_controller {
```
```suggestion
           tmc50xx_1_motion_controller: motion-controller-1 {
```

JarmouniA commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -656,43 +656,40 @@ Stepper
+  * :c:func:`stepper_stop` is replaced by :c:func:`stepper_ctrl_stop`.
+  * :c:func:`stepper_set_reference_position` is replaced by :c:func:`stepper_ctrl_set_reference_position`.
+  * :c:func:`stepper_get_actual_position` is replaced by :c:func:`stepper_ctrl_get_actual_position`.
+  * :c:func:`stepper_set_microstep_interval` is replaced by :c:func:`stepper_ctrl_set_microstep_interval`.
```
```suggestion
* Several API functions have been renamed:

  * ``stepper_move_by`` to :c:func:`stepper_ctrl_move_by`.
  * ``stepper_move_to`` to :c:func:`stepper_ctrl_move_to`.
  * ``stepper_is_moving`` to :c:func:`stepper_ctrl_is_moving`.
  * ``stepper_run`` to :c:func:`stepper_ctrl_run`.
  * ``stepper_stop`` to :c:func:`stepper_ctrl_stop`.
  * ``stepper_set_reference_position`` to :c:func:`stepper_ctrl_set_reference_position`.
  * ``stepper_get_actual_position`` to :c:func:`stepper_ctrl_get_actual_position`.
  * ``stepper_set_microstep_interval`` to :c:func:`stepper_ctrl_set_microstep_interval`.
```

JarmouniA commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -656,43 +656,40 @@ Stepper
+
+* ``step-gpios``, ``dir-gpios``, ``invert-direction`` and ``counter`` properties are removed
+  from :dtcompatible:`adi,tmc2209`, :dtcompatible:`ti,drv84xx` and :dtcompatible:`allegro,a4979`,
+  these are now are implemented by :dtcompatible:`zephyr,gpio-step-dir-stepper-ctrl`.
```
```suggestion
* ``step-gpios``, ``dir-gpios``, ``invert-direction`` and ``counter`` properties have been removed
  from devicetree bindings :dtcompatible:`adi,tmc2209`, :dtcompatible:`ti,drv84xx` and :dtcompatible:`allegro,a4979`,
  and moved to :dtcompatible:`zephyr,gpio-step-dir-stepper-ctrl` binding.
```

JarmouniA commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -656,43 +656,40 @@ Stepper
+* :c:func:`stepper_ctrl_move_by`, :c:func:`stepper_ctrl_move_to`, :c:func:`stepper_ctrl_run`,
+  :c:func:`stepper_ctrl_stop`, :c:func:`stepper_ctrl_is_moving`, :c:func:`stepper_ctrl_set_microstep_interval`
+  and :c:func:`stepper_ctrl_set_event_cb` APIs are not implemented by :dtcompatible:`adi,tmc2209`,
+  :dtcompatible:`ti,drv84xx` and :dtcompatible:`allegro,a4979`.
```
What does this even mean? Were their software drivers implementing these API functions and now they are not? Are these functions optional?
The migration guide should only be talking about stuff affecting user apps (changes in public APIs, devicetree, and visible Kconfig options), and what the user has to do to get their app working as before.

JarmouniA commented on samples/drivers/stepper/generic/boards/nucleo_g071rb.overlay:
```diff
@@ -18,11 +18,11 @@
 
-	step_dir_motion_control: step_dir_motion_control {
-		compatible = "zephyr,gpio-step-dir-stepper";
+	step_dir_motion_control: step-dir-motion-control {
```
```suggestion
	step_dir_motion_control: motion-controller {
```
https://devicetree-specification.readthedocs.io/en/latest/chapter2-devicetree-basics.html#generic-names-recommendation

jilaypandya commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -656,43 +656,40 @@ Stepper
+  * :c:func:`stepper_get_actual_position` is replaced by :c:func:`stepper_ctrl_get_actual_position`.
+  * :c:func:`stepper_set_microstep_interval` is replaced by :c:func:`stepper_ctrl_set_microstep_interval`.
+
+* The following events have been moved from :c:enum:`stepper_event` to :c:enum:`stepper_ctrl_event`:
```
The original stepper api was a monolithic api comprising of both motion controller and driver functions.
We have recently split that monolith into two seperate device driver classes, as a part of which some events have been moved from `stepper_event` enum class to `stepper_ctrl_event` enum class. The original `stepper_event` class still remains in the codebase, however only contains stepper hardware driver events now.

JarmouniA commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -656,43 +656,56 @@ Stepper
+* Motion control must now be performed through a :dtcompatible:`zephyr,gpio-step-dir-stepper-ctrl`
+  device that references the stepper driver via the ``stepper-driver`` property. Applications must
+  update their devicetree to add a motion controller node and use the ``stepper_ctrl_*`` APIs
+  instead of calling motion control functions directly on the driver device.
```
```suggestion
  device that references the stepper hardware driver devicetree node via the ``stepper-driver`` property. Applications must
  update their devicetree to add a motion controller node and use the ``stepper_ctrl_*`` APIs,
  instead of calling motion control functions directly on the stepper hardware driver device.
```

JarmouniA commented on dts/bindings/stepper/adi/adi,tmc50xx.yaml:
```diff
@@ -55,17 +55,17 @@ examples:
             };
 
-            tmc50xx_0_motion_controller: tmc50xx_0_motion_controller {
+            tmc50xx_0_motion_controller: tmc50xx-0-motion-controller {
```
```suggestion
            tmc50xx_0_motion_controller: motion-controller-0 {
```

JarmouniA commented on include/zephyr/drivers/stepper/stepper.h:
```diff
@@ -0,0 +1,276 @@
+/**
+ * @brief Stepper Hardware Driver API
+ */
+__subsystem struct stepper_driver_api {
```
```suggestion
__subsystem struct stepper_hw_driver_api {
```

jbehrensnx reviewed (changes requested):
One minor blocking change.

Some things that I noticed and that need to be fixed, but that can be done in followup PRs:

The year numbers in many Copyright entries need to be updated. Considering how pervasive this issue seems, a separate PR is probably best for this. 

In addition, there a quite a few standalone `stepper` entries in the documentation of code files. They should be updated to clarify which component they mean (which is not always obvious from the file they are in). As this was already an issue before this PR, one can consider updating them out of scope for this PR.

JarmouniA commented on doc/hardware/peripherals/stepper/index.rst:
```diff
@@ -5,39 +5,39 @@ Steppers
 ***********
 
-In the context of stepper controllers  device tree provides the initial hardware
+In the context of stepper motion controllers  devicetree provides the initial hardware
```
the correct thing is `In the context of stepper motion controllers, devicetree...`
